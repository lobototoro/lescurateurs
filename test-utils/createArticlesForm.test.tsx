import CreateArticleForm from "@/app/editor/components/formComponents/createArticlesForm";
import { cleanup, fireEvent, render, waitFor } from "@testing-library/react";
import { expect, it, vi, describe, afterEach } from "vitest";

vi.mock('@/app/articleActions', () => ({
  createArticleAction: vi.fn().mockImplementation(() => {
    return {
      message: true,
      text: 'Article created'
    }
  })
}));

vi.mock('next/font/google', () => ({
  Alegreya: () => ({
    style: {
      fontFamily: 'mocked',
    },
  }),
  Raleway: () => ({
    style: {
      fontFamily: 'mocked',
    },
  })
}));

describe('Create article form', () => {
  afterEach(() => {
    cleanup();
  });

  it('should render correctly', () => {
    const { asFragment } = render(<CreateArticleForm />);
    
    expect(asFragment()).toMatchSnapshot();
  });

  it('should return an ok response after creation', async () => {
    const { getByText, getByTestId } = render(<CreateArticleForm />);

    fireEvent.change(getByTestId('title'), {
      target: { value: 'My title' }
    });
    
    fireEvent.change(getByTestId('introduction'), {
      target: { value: 'Nisi enim dolor quis in ullamco laboris. Amet nulla adipisicing irure minim mollit excepteur deserunt anim deserunt nulla dolore.' }
    });

    fireEvent.change(getByTestId('main'), {
      target: { value: 'Nisi enim dolor quis in ullamco laboris. Amet nulla adipisicing irure minim mollit excepteur deserunt anim deserunt nulla anim non incididunt. Occaecat culpa velit irure mollit quis occaecat nulla. Nostrud cupidatat do sunt in do. Sint consectetur tempor nulla est ut dolor incididunt proident sunt aliquip nisi do aute voluptate.' }
    });

    fireEvent.change(getByTestId('mainAudioUrl'), {
      target: { value: 'https://example.com' }
    });

    fireEvent.change(getByTestId('urlToMainIllustration'), {
      target: { value: 'https://example.com' }
    });

    fireEvent.click(getByTestId('submit'));

    await waitFor(() => {
      expect(getByText('Article created')).toBeDefined();
    });
  });

  it("should return an error response after creation", async () => {
    const { getByTestId } = render(<CreateArticleForm />);

    fireEvent.change(getByTestId('title'), {
      target: { value: 'My title' }
    });
    
    fireEvent.change(getByTestId('introduction'), {    
      target: { value: 'Nisi enim dolor quis in ullamco laboris. Amet nulla adipisicing irure minim mollit excepteur deserunt anim deserunt nulla dolore.' }
    });  

    fireEvent.change(getByTestId('main'), {
      target: { value: 'Nisi enim dolor quis in ullamco laboris. Amet nulla adipisicing irure minim mollit excepteur deserunt anim deserunt nulla anim non incididunt. Occaecat culpa velit irure mollit quis occaecat nulla. Nostrud cupidatat do sunt in do. Sint consectetur tempor nulla est ut dolor incididunt proident sunt aliquip nisi do aute voluptate.' }
    });

    fireEvent.change(getByTestId('mainAudioUrl'), {
      target: { value: 'my url' }
    });

    fireEvent.change(getByTestId('urlToMainIllustration'), {
      target: { value: 'https://example.com' }
    });

    fireEvent.click(getByTestId('submit'));

    await waitFor(() => {
      const modal = getByTestId('create-article-modal');
      const modalInnerContent = modal.querySelector('.modal-content');
      expect(modalInnerContent?.classList.contains('is-danger')).toBeTruthy();
    });
  });
});