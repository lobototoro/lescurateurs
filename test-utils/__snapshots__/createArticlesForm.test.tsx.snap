// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Create article form > should render correctly 1`] = `
<DocumentFragment>
  <form>
    <div
      class="mt-6"
    >
      <div
        class="field"
      >
        <label
          aria-label="label du champ Titre"
          class="label"
          for="title"
        >
          Titre
        </label>
        <div
          class="field"
        >
          <input
            class="input"
            data-testid="title"
            name="title"
            required=""
            type="text"
          />
        </div>
      </div>
      <div
        class="field"
      >
        <label
          aria-label="label du champ introduction"
          class="label"
          for="introduction"
        >
          introduction
        </label>
        <textarea
          class="textarea"
          data-testid="introduction"
          name="introduction"
          required=""
          rows="5"
        />
      </div>
      <div
        class="field"
      >
        <label
          aria-label="label du champ Texte"
          class="label"
          for="main"
        >
          Texte
        </label>
        <textarea
          class="textarea"
          data-testid="main"
          name="main"
          required=""
          rows="10"
        />
      </div>
      <div
        class="field"
      >
        <label
          aria-label="label du champ lien vers l'illustration"
          class="label is-inline-flex"
          for="urlToMainIllustration"
        >
          lien vers l'illustration
        </label>
        <div
          class="control"
        >
          <input
            class="input"
            data-testid="urlToMainIllustration"
            name="urlToMainIllustration"
            required=""
            type="url"
          />
        </div>
      </div>
      <div
        class="field"
      >
        <label
          aria-label="label du champ lien audio principal"
          class="label"
          for="mainAudioUrl"
        >
          lien audio principal
        </label>
        <div
          class="control"
        >
          <input
            class="input"
            data-testid="mainAudioUrl"
            name="mainAudioUrl"
            required=""
            type="url"
          />
        </div>
      </div>
      <div>
        <h4
          class="undefined article-title_medium article-title_white mt-5 mb-4"
        >
          Ajouter des liens multim√©dias
        </h4>
      </div>
      <div
        class="field"
      >
        <div
          class="mt-5"
        >
          <div
            class="button is-primary is-size-6 has-text-white"
          >
            +
          </div>
          <div
            class="button is-primary ml-2 is-size-6 has-text-white"
          >
            -
          </div>
        </div>
      </div>
    </div>
    <div
      class="modal"
      data-testid="create-article-modal"
    >
      <div
        class="modal-background"
      />
      <div
        class="modal-content is-danger"
      >
        <p
          class="is-size-6 has-text-white has-background-primary p-6"
        />
        <footer>
          <button
            aria-label="accept button"
            class="button"
          >
            OK
          </button>
        </footer>
      </div>
    </div>
    <div
      class="field mt-5"
    >
      <input
        class="button is-primary is-size-6 has-text-white"
        data-testid="submit"
        type="submit"
        value="Valider"
      />
    </div>
  </form>
</DocumentFragment>
`;
